### ISSUE TITLE
Excessive small object creation/deletion impacting overall performance

### ISSUE THREAD

**Issue by mmd-osm** [drolbr/Overpass-API#196]
**TL;DR:**
collect_items creates a huge amount of mostly unneeded small objects (=performance implications).

--

Performance measurements revealed an overall performance penalty in the way Overpass API uses block_backend in Template DB. More specifically, quite a huge number of very small objects is allocated and deallocated over the course of a query execution. E.g. the example query 1 below created 6'810'334'168 small objects. 

To get a first indication if we're affected by excessive object creation, I did some measurements using a different malloc implementation, namely [Google's tcmalloc](http://goog-perftools.sourceforge.net/doc/tcmalloc.html). By ld_preloading this library, a different malloc/free implemention is used instead of the standard one in libstd++. I could reduce the overall runtime across a variety of queries by 20-25% this way.

This however does not change the overall number of malloc/free operations, which I believe is the main driver for the performance impact. By addressing excessive malloc/free ops, the potential improvements are likely to be higher and will - to a certain degree - affect any query reading data from the database.

I will focus on collect_items_range to illustrate the issue, although other parts are likely affected in a similar way.

```C
void collect_items_range
{
  for ( it(db.range_begin(req.begin(), req.end()));
	   !(it == db.range_end()); ++it)
  {
    if (predicate.match(it.object()))
      result[it.index()].push_back(it.object());
  }
}
```

Here, a predicate is used to determine if an it.object() is in the result list. However, the Block_Backend nevertheless has to create object instances for a huge number of entries, most of which have to be discarded immediately again.

Right now, it appears that `Block_Backend_Basic_Iterator< TIndex, TObject >::object()` creates full blown object instances (like Node_Skeleton), while `Id_Predicate` only checks for the `id`.

To avoid unneeded object instantiation, some new methods will be required to extract an object id from a raw buffer and apply the predicate on this object id. Only in case of a match, a new object is created like today.

**Example Query 1:** http://overpass-turbo.eu/s/7Gz

[jemalloc analysis](https://github.com/jemalloc/jemalloc/wiki/Use-Case%3A-Basic-Allocator-Statistics) (showing number of object allocations by object size amongst others)

```
                            allocated      nmalloc      ndalloc    nrequests
small:                         135464    166594686    166593915   6810334168
large:                          77824      6095392      6095390      8441719
huge:                               0            0            0            0
total:                         213288    172690078    172689305   6818775887
active:                        569344
mapped:                      33554432
```
=> 6.8 billion objects created/deleted

**Example query 2:** worldwide pharmacy POIs:

```C
[timeout:3600][out:json];
(node[amenity=pharmacy];
way[amenity=pharmacy];
rel[amenity=pharmacy]);
out count;
```

```
dirty pages: 137:989 active:dirty, 149 sweeps, 177 madvises, 11813 purged
                            allocated      nmalloc      ndalloc    nrequests
small:                         126136      1302972      1302305   1134243840
large:                          77824          391          389          630
huge:                               0            0            0            0
total:                         203960      1303363      1302694   1134244470
```

=> 1 billion objects need to be created/deleted (!)

**Comment by mmd-osm** [comment]
The following changes are based on a slightly different approach: instead of pushing down the predicate, we're pulling the raw data from the Block Backend, determine the `id` without any additional object instantiation and pass that id to an adapted `match` method.

Depending on the test case, theses changes reduces the user time by 17-45%.

struct Node_Skeleton
```C
  static Node::Id_Type get_id(const void* data)
  {
    return *(uint64*)data;
  }
```

struct Way_Skeleton
```C
  static Way::Id_Type get_id(const void* data)
  {
    return *(uint32*)data;
  }
```

struct Relation_Skeleton
```C
  static Relation::Id_Type get_id(const void* data)
  {
    return *(uint32*)data;
  }
```

struct Area_Skeleton
```C
  static Area::Id_Type get_id(const void* data)
  {
    return *(uint32*)data;
  }
```

template < class TObject >
class Id_Predicate
```C
    bool match(const typename TObject::Id_Type id) const
    {
      return binary_search(ids.begin(), ids.end(), id);
    }
```

template< class TIndex, class TObject >
struct Block_Backend_Basic_Iterator:
```C
  const void* object_data();


template< class TIndex, class TObject >
const void * Block_Backend_Basic_Iterator< TIndex, TObject >::object_data()
{
  return ((void*)(buffer.ptr + pos));
}
```

void collect_items_range
```C
    //if (predicate.match(it.object())) {
    if (predicate.match(Object::get_id(it.object_data()))) {
      result[it.index()].push_back(it.object());
    }
```

**Comment by mmd-osm** [comment]
Seems to be a common pattern: another location with a similar issue: print.cc: void collect_tags_framed:

```C
    if (!(tag_it.object() < lower_id_bound) &&
      (tag_it.object() < upper_id_bound) &&
      (binary_search(ids_by_coarse[coarse_index].begin(),
```

Instead of pulling an Uint64 value from the buffer.ptr and passing the Uint64 bit value in a register, a new object is created ... 

```
#0  0x00007f09201f72e0 in malloc () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007f0920ab4f2d in operator new(unsigned long) ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#2  0x00000000004e55ba in object (this=0x7fffa1ac4cc0)
    at ../../src/test-bin/../overpass_api/statements/../data/../statements/../dispatch/../osm-backend/../../template_db/block_backend.h:346
#3  collect_tags_framed<Uint64> (tags_by_id=..., items_db=..., tag_it=..., 
    ids_by_coarse=..., coarse_index=1646680064, lower_id_bound=..., 
    lower_id_bound@entry=..., upper_id_bound=upper_id_bound@entry=...)
    at ../../src/test-bin/../overpass_api/statements/print.cc:543
#4  0x00000000004ead81 in Print_Statement::tags_by_id<Uint32_Index, Node_Skeleton> (this=this@entry=0x1add340, items=..., file_prop=..., 
    FLUSH_SIZE=FLUSH_SIZE@entry=1048576, target=..., rman=..., 
    transaction=..., meta_file_prop=meta_file_prop@entry=0x0, 
    element_count=@0x7fffa1ac4fe0: 0)
    at ../../src/test-bin/../overpass_api/statements/print.cc:923
#5  0x00000000004d39e8 in Print_Statement::execute (this=0x1add340, rman=...)
    at ../../src/test-bin/../overpass_api/statements/print.cc:2353
```

Not quite sure what the best general approach could look like. Extracting single 32bit or 64bit values  from buffer.ptr (from block backend) is a quite common. Creating dedicated objects seems be a bit heavy in hot code paths.

**Comment by mmd-osm** [comment]
Also see https://github.com/drolbr/Overpass-API/pull/208/files#r64969784

**Comment by drolbr** [comment]
Fixed in commit 46f8fb93d5f214971b6be8233ebe082e7ecf2e3d .

The provided pull request does solve the problem. However, it patches a part of the code that is already rather confusing. The implemented solution should make this part of the code easier to understand.
